### PR TITLE
[SPORTEC] Resolved Referee Issue

### PR DESCRIPTION
**Pull request by UnravelSports**
Hi!

When loading the DFL open data with the new functionality (introduced in https://github.com/PySport/kloppy/pull/365) two match `J03WN1` and `J03WMX` had issues loading. The simple reason for this was the introduction of the `TeamId="referee"`. This lead to unknown player Id `PersonId="DFL-OBJ-00011X"` because the referee Ids would not be parsed from the meta data, and were/are not present in the `player_map` in the deserializer.

```xml
<FrameSet GameSection="firstHalf" MatchId="DFL-MAT-J03WN1" TeamId="referee" PersonId="DFL-OBJ-00011X">
<Frame N="10000" T="2023-05-27T13:31:06.400+00:00" X="3.06" Y="-11.56" D="0.00" S="0.00" A="0.00" M="1"/>
```

The below in line [#188](https://github.com/PySport/kloppy/blob/7d3c4580a8f103e10a2301e67593bbaa3e9e1c6b/kloppy/infra/serializers/tracking/sportec/deserializer.py#L188) of the Deserializer would throw a KeyError.
```
player_map[player_id]: PlayerData(
      coordinates=Point(
          x=float(raw_player_data["X"]),
          y=float(raw_player_data["Y"]),
      ),
      speed=float(raw_player_data["S"]),
  )
```

I wanted a clean way to resolve this, and include some potential future functionality that would allow us to include the referee coordinates into the tracking dataset at some point. 

This means I have introduced a simple `RefereeType`:

```python
class RefereeType:
    VideoReferee = "Video Referee"
    Referee = "Referee"
    Assistant = "Assistant"
    FourthOfficial = "Fourth Official"
```

I added this to the `SportecMetadata` class:

```python
class SportecMetadata(NamedTuple):
    score: Score
    teams: List[Team]
    periods: List[Period]
    x_max: float
    y_max: float
    fps: int
    home_coach: str
    away_coach: str
    referees: List[Referee]
```

And we parse the referees into this, if they exist:

```python
if hasattr(match_root, "MatchInformation") and hasattr(
        match_root.MatchInformation, "Referees"
    ):
        referees = []
        referee_path = objectify.ObjectPath(
            "PutDataRequest.MatchInformation.Referees"
        )
        referee_elms = referee_path.find(match_root).iterchildren(
            tag="Referee"
        )

        for referee in referee_elms:
            ref_attrib = referee.attrib
            referees.append(
                Referee(
                    referee_id=ref_attrib["PersonId"],
                    name=ref_attrib["Shortname"],
                    first_name=ref_attrib["FirstName"],
                    last_name=ref_attrib["LastName"],
                    role=referee_types_mapping[ref_attrib["Role"]],
                )
            )
    else:
        referees = []

    return SportecMetadata(
        score=score,
        teams=teams,
        periods=periods,
        x_max=x_max,
        y_max=y_max,
        fps=SPORTEC_FPS,
        home_coach=home_coach,
        away_coach=away_coach,
        referees=referees,
    )
```

Here the referee mapping refers the below:
```python
referee_types_mapping: Dict[str, RefereeType] = {
    "referee": RefereeType.Referee,
    "firstAssistant": RefereeType.Assistant,
    "videoReferee": RefereeType.VideoReferee,
    "secondAssistant": RefereeType.Assistant,
    "fourthOfficial": RefereeType.FourthOfficial,
}
```

Then in the actual deserializer we simply check if any of the player_ids is not a `referee_id` (along side checking if player_id is not "ball").

This, for now, simply skips the parsing of the referee coordinates.

Finally, I have added the below to the overall `Metadata` class, and we now add this to the Metadata when we parse the Sportec tracking data (and the event data too, since we have this functionality now anyway)

```python
referees: Optional[List] = field(default_factory=list)
```

```python
metadata = Metadata(
            teams=teams,
            periods=periods,
            pitch_dimensions=transformer.get_to_coordinate_system().pitch_dimensions,
            score=sportec_metadata.score,
            frame_rate=sportec_metadata.fps,
            orientation=orientation,
            provider=Provider.SPORTEC,
            flags=DatasetFlag.BALL_OWNING_TEAM | DatasetFlag.BALL_STATE,
            coordinate_system=transformer.get_to_coordinate_system(),
            date=date,
            game_week=game_week,
            game_id=game_id,
            home_coach=home_coach,
            away_coach=away_coach,
            referees=sportec_metadata.referees,
        )
```

Finally, this allows us to test if we have parsed the referees successfully in a new test, and in the test_load_metadata test. For this test I've introduced a new file `kloppy/tests/files/sportec_positional_w_referee.xml` which simply includes coordinates as below.

For now we simply care that the test works (so we skipped parsing the referees) and that the referees are returned successfully.

```xml
<FrameSet GameSection="secondHalf" MatchId="DFL-MAT-003BN1" TeamId="referee" PersonId="DFL-OBJ-000123">
      <Frame N="100000" T="2020-06-05T21:33:24.920+02:00" X="-3.91" Y="14.10" S="0.00" D="0.00" A="0.00" M="46"/>
```

```python
def test_referees(self, raw_data_referee: Path, meta_data: Path):
        dataset = sportec.load_tracking(
            raw_data=raw_data_referee,
            meta_data=meta_data,
            coordinates="sportec",
            only_alive=True,
        )
        assert len(dataset.metadata.referees) == 4
```

### CODE REVIEW

File: kloppy/domain/models/common.py [hunk 1]
```diff
@@ -119,6 +119,25 @@ def __str__(self):
         return self.value
 
 
+@dataclass(frozen=True)
+class Referee:
+    referee_id: str
+    name: str = None
+    first_name: str = None
+    last_name: str = None
```
Comment (probberechts) [inline kloppy/domain/models/common.py:127]:
```suggestion
    name: Optional[str] = None
    first_name: Optional[str] = None
    last_name: Optional[str] = None
```

File: kloppy/domain/models/common.py [hunk 2]
```diff
@@ -119,6 +119,25 @@ def __str__(self):
...
+
+    @property
+    def full_name(self):
+        if self.name:
+            return self.name
+        if self.first_name or self.last_name:
+            return f"{self.first_name} {self.last_name}"
+        if self.role:
+            return f"{self.role}_{self.referee_id}"
+        return self.referee_id
```
Comment (probberechts) [inline kloppy/domain/models/common.py:138]: For players we use `home_{jersey_no}` or `away_{jersey_no}` as the final fallback. This makes it still clear that they are players. Maybe we hava to add "referee" as a prefix here.

```suggestion
        return f"referee_{self.referee_id}"
```

File: kloppy/domain/models/common.py [hunk 3]
```diff
@@ -119,6 +119,25 @@ def __str__(self):
...
+    role: Optional[RefereeType] = None
+
+    @property
+    def full_name(self):
+        if self.name:
+            return self.name
+        if self.first_name or self.last_name:
+            return f"{self.first_name} {self.last_name}"
+        if self.role:
+            return f"{self.role}_{self.referee_id}"
```
Comment (probberechts) [inline kloppy/domain/models/common.py:137]: For players, we use a lowercase string if the name is unknown, like `home_{jersey_no}` or `away_{jersey_no}`. For consistency, I would use lowercase for the role of the referee too.

File: kloppy/domain/models/common.py [hunk 4]
```diff
@@ -119,6 +119,25 @@ def __str__(self):
         return self.value
 
 
+@dataclass(frozen=True)
+class Referee:
```
Comment (probberechts) [inline kloppy/domain/models/common.py:123]: "Referee" or "Official"? I don't have a strong preference but one reason to prefer "Official" covers (e.g., the fourth official).

File: kloppy/domain/models/position.py [hunk 5]
```diff
@@ -92,3 +92,10 @@ def __str__(self):
     @classmethod
     def unknown(cls) -> "PositionType":
         return cls.Unknown
+
+
+class RefereeType:
+    VideoReferee = "Video Referee"
+    Referee = "Referee"
```
Comment (probberechts) [inline kloppy/domain/models/position.py:99]: They are all a referee, hence using `RefereeType="Referee"` seems ambiguous.

```suggestion
    Referee = "Main Referee"
```

File: kloppy/domain/models/position.py [hunk 6]
```diff
@@ -92,3 +92,10 @@ def __str__(self):
     @classmethod
     def unknown(cls) -> "PositionType":
         return cls.Unknown
+
+
+class RefereeType:
+    VideoReferee = "Video Referee"
```
Comment (probberechts) [inline kloppy/domain/models/position.py:98]:
```suggestion
    VideoReferee = "Video Assistant Referee"
```

File: kloppy/domain/models/position.py [hunk 7]
```diff
@@ -92,3 +92,10 @@ def __str__(self):
...
+    Referee = "Referee"
+    Assistant = "Assistant"
```
Comment (probberechts) [inline kloppy/domain/models/position.py:100]:
```suggestion
    Assistant = "Assistant Referee"
```

File: kloppy/domain/models/common.py (hunk 4 above)
Comment (UnravelSports) [reply kloppy/domain/models/common.py:123]: I wasn't sure here either, but I will update it to Official.

**Comment by UnravelSports** [comment]
@probberechts I have made all suggested changes and converted "Referee" and "RefereeType" to "Official" and "OfficialType" respectively.

**Pull request merged**